exchanges/okgroup/okgroup_websocket.go

```diff
@@ -0,0 +1,759 @@
+		o.GetName(), event.ErrorCode, event.Message)
+	if o.Verbose {
+		log.Error(errorMessage)
+	}
+	o.Websocket.DataHandler <- fmt.Errorf(errorMessage)
+}
+
+// GetWsChannelWithoutOrderType takes WebsocketDataResponse.Table and returns
+// The base channel name eg receive "spot/depth5:BTC-USDT" return "depth5"
+func (o *OKGroup) GetWsChannelWithoutOrderType(table string) string {
+	index := strings.Index(table, "/")
+	if index == -1 {
+		return table
+	}
+	channel := table[index+1:]
+	index = strings.Index(channel, ":")
+	// Some events do not contain a currency
+	if index == -1 {
+		return channel
+	}
+
+	return channel[:index]
+}
+
+// GetAssetTypeFromTableName gets the asset type from the table name
+// eg "spot/ticker:BTCUSD" results in "SPOT"
+func (o *OKGroup) GetAssetTypeFromTableName(table string) string {
+	assetIndex := strings.Index(table, "/")
+	return strings.ToUpper(table[:assetIndex])
+}
+
+// WsHandleDataResponse classifies the WS response and sends to appropriate handler
+func (o *OKGroup) WsHandleDataResponse(response *WebsocketDataResponse) {
+	switch o.GetWsChannelWithoutOrderType(response.Table) {
+	case okGroupWsCandle60s, okGroupWsCandle180s, okGroupWsCandle300s, okGroupWsCandle900s,
+		okGroupWsCandle1800s, okGroupWsCandle3600s, okGroupWsCandle7200s, okGroupWsCandle14400s,
+		okGroupWsCandle21600s, okGroupWsCandle43200s, okGroupWsCandle86400s, okGroupWsCandle604900s:
+		if o.Verbose {
+			log.Debugf("%v Websocket candle data received", o.GetName())
+		}
+		o.wsProcessCandles(response)
+	case okGroupWsDepth, okGroupWsDepth5:
+		if o.Verbose {
+			log.Debugf("%v Websocket orderbook data received", o.GetName())
+		}
+		// Locking, orderbooks cannot be processed out of order
+		orderbookMutex.Lock()
+		err := o.WsProcessOrderBook(response)
+		if err != nil {
+			log.Error(err)
+			subscriptionChannel := fmt.Sprintf("%v:%v", response.Table, response.Data[0].InstrumentID)
+			o.ResubscribeToChannel(subscriptionChannel)
+		}
+		orderbookMutex.Unlock()
+	case okGroupWsTicker:
+		if o.Verbose {
+			log.Debugf("%v Websocket ticker data received", o.GetName())
+		}
+		o.wsProcessTickers(response)
+	case okGroupWsTrade:
+		if o.Verbose {
+			log.Debugf("%v Websocket trade data received", o.GetName())
+		}
+		o.wsProcessTrades(response)
+	default:
+		logDataResponse(response)
+	}
+}
+
+// resubscribeToChannel will attempt to unsubscribe and resubscribe to a channel
+func (o *OKGroup) ResubscribeToChannel(channel string) {
+	if okGroupWsResubscribeFailureLimit > 0 {
+		var successfulUnsubscribe bool
+		for i := 0; i < okGroupWsResubscribeFailureLimit; i++ {
+			err := o.WsUnsubscribeToChannel(channel)
+			if err != nil {
+				log.Error(err)
+				time.Sleep(okGroupWsResubscribeDelayInSeconds * time.Second)
+				continue
+			}
+			successfulUnsubscribe = true
+			break
+		}
+		if !successfulUnsubscribe {
+			log.Fatalf("%v websocket channel %v failed to unsubscribe after %v attempts", o.GetName(), channel, okGroupWsResubscribeFailureLimit)
```

Comment:
This will be improved in a seperate PR which introduces a websocket client which keeps track of all subscribed/unsubcribed channels as well as initialising an exchanges WS DataHandler but for now this will suffice.